pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,139 @@
+	"planetscale.dev/vitess-operator/pkg/operator/vttablet"
+)
+
+func (r *ReconcileVitessShard) reconcileRollout(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
+	if !rollout.Cascaded(vts) {
+		// If the shard is not scheduled for a cascading update, silently bail out and do nothing.
+		return resultBuilder.Result()
+	}
+
+	podList := &v1.PodList{}
+	listOpts := &client.ListOptions{
+		Namespace: vts.Namespace,
+		LabelSelector: apilabels.Set(map[string]string{
+			planetscalev2.ComponentLabel: planetscalev2.VttabletComponentName,
+			planetscalev2.ClusterLabel:   vts.Labels[planetscalev2.ClusterLabel],
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	// Safety checks and rolling updates must be deterministically ordered, so we access and sort pods by tablet alias.
+	tabletPods := make(map[string]v1.Pod)
+	for i := range podList.Items {
+		pod := podList.Items[i]
+		tabletAlias := vttablet.AliasFromPod(&pod)
+		tabletKey := topoproto.TabletAliasString(&tabletAlias)
+		tabletPods[tabletKey] = pod
+	}
+
+	tabletKeys := make([]string, 0, len(tabletPods))
+	for key := range tabletPods {
+		tabletKeys = append(tabletKeys, key)
+	}
+	sort.Strings(tabletKeys)
+
+	for _, tabletKey := range tabletKeys {
+		tablet := vts.Status.Tablets[tabletKey]
+		if tablet.Available != v1.ConditionTrue {
+			// If all tablets aren't healthy, we should bail and not perform a rolling restart.
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for tablet %v to be Available.", tabletKey)
+			return resultBuilder.Result()
+		}
+
+		pod := tabletPods[tabletKey]
+		if rollout.Released(&pod) {
+			// If any tablet has already been released, we should wait until it is finished to release another one.
+			r.recorder.Eventf(vts, v1.EventTypeNormal, "RolloutPaused", "Waiting for tablet %v to finish release.", tabletKey)
+		}
+	}
+
+	scheduledTablets := false
+	masterEligibleTablets := vts.Spec.MasterEligibleTabletCount()
+
+	for _, tabletKey := range tabletKeys {
+		pod := tabletPods[tabletKey]
+		if !rollout.Scheduled(&pod) {
+			continue
+		}
+
+		scheduledTablets = true
+		deletePod := false
+		tabletType := pod.Labels[planetscalev2.TabletTypeLabel]
+		// These two conditions guarantee that the tablet is a lone master.
+		if masterEligibleTablets == 1 && tabletType == string(planetscalev2.ReplicaPoolType) {
+			// If we have a lone master, we must delete it since reparenting is impossible.
+			deletePod = true
+		}
+
+		if err := r.releaseTabletPod(ctx, &pod, deletePod); err != nil {
+			r.recorder.Eventf(vts, v1.EventTypeWarning, "RollingRestartFailed", "deletion of pod %v with tablet %v failed", pod.Name, tabletKey)
+			return resultBuilder.Error(err)
```

Comment:
I'm not sure if it makes sense to fail the entire process here. I would defer to @enisoc 's judgment on this one. Potential suggestion: 
```suggestion
			resultBuilder.Error(err)
```

And then check for error after the loop and bail then so we at least get through as many pods as possible before a retry.

Reply:
`reconcileRollout` will at most release one Pod for a rolling update no matter what. Even if we don't error out here, we would release only the next scheduled pod. This might not be worth losing our deterministic rollout, but I would also defer to @enisoc's judgment.

Reply:
> check for error after the loop and bail then so we at least get through as many pods as possible

We actually don't want to get through any other Pods in this case. We are looking for the one Pod whose turn is up next, and then we try to process that one Pod.

Maybe we should refactor the code a bit to enforce this more clearly: instead of doing this inside a loop, make a helper like `getFirstScheduledTabletKey` which returns a single tabletKey, or `""` if no tablets are scheduled. Then it's clear that we only ever look at one Pod per reconcile.

It also makes it easier to treat `uncascadeShard` as a bail-out so the real meat of the function (releasing the next tablet) is less indented.

Reply:
Ahhh yeah, then I would say the code could be a bit more clear with a helper like that which filters first. As a reader this came across as if we were releasing more than one pod.